packages/client/src/runtime/RequestHandler.ts

```diff
@@ -96,28 +99,30 @@ export class RequestHandler {
         const queries = requests.map((r) => r.protocolMessage.toEngineQuery())
         const traceparent = getTraceParent({ context: requests[0].otelParentCtx, tracingConfig: client._tracingConfig })
 
-        if (traceparent) info.headers.traceparent = traceparent
+        if (traceparent) info.traceparent = traceparent
         // TODO: pass the child information to QE for it to issue links to queries
         // const links = requests.map((r) => trace.getSpanContext(r.otelChildCtx!))
 
         const containsWrite = requests.some((r) => r.protocolMessage.isWrite())
 
-        const batchTransaction = info.transaction?.kind === 'batch' ? info.transaction : undefined
-
         return this.client._engine.requestBatch(queries, {
-          headers: info.headers,
-          transaction: batchTransaction,
+          traceparent,
+          transaction: getTransactionOptions(info.transaction),
           containsWrite,
+          customDataProxyHeaders: info.customDataProxyHeaders,
         })
       },
       singleLoader: (request) => {
         const info = getRequestInfo(request)
-        const interactiveTransaction = info.transaction?.kind === 'itx' ? info.transaction : undefined
+        const interactiveTransaction =
```

Comment:
I just noticed that this may be wrong. It is possible for a batch transaction to have a single query. Thay feels edge-casey but it may be useful to set an isolation level of a raw query, for instance. So my suggestion is to just pass the transaction options here as well.

Reply:
Nicely spot, i think that was broken since forever. Fix is actually more complicated than what you suggesting - only `requestBatch` has transaction options, not singular `requests`. So, if it's a single request with batch tx options set, it has to go through `requestBatch` endpoint. This what I did in separate pr #17556 since it is a separate issue.